### DISCUSSION TITLE
Shallow copies of Session instances to promote composition?

### DISCUSSION THREAD

**Discussion by ursusursus** [Alamofire/Alamofire#3627] [General]
I have a case where I have multiple logical levels of network apis - pubic, semi-public, private, where private inherits semi-public, and semi-public inherits from public. 

Each adds its own interceptors to attach headers and other behavior and should inherit configuration of the parent.
In OkHttp in java/kotlin world this is done via creating a shallow copy on the okhttclient instance

```
val semiPublicOkHttpClient = publicOkHttpClient
    .newBuilder()
    ...// attach semi public interceptors
    .build()
``` 

Since its a shallow copy, threadpools are not duplicated, so its not a performance/memory issue.

If I setup instances this way, then they can be simply injected into feature api interface and get the "feature" developer gets all the desired shared behavior for free, which is great.

So, mostly a question, since I dont believe this is supported. I looked at SessionDelegate which sounded what I want but doesnt seem to do so.

Best api would look like this
```
let publicSession = Session(interceptors: publicInterceptors] // Notice no parent
let semiPublicSession = Session(parent: publicSession, interceptors: semiPublicInterceptors]
let privateSession = Session(parent: semiPublicSession, interceptors: privateInterceptors]
```

I can sort of emulate this my self by collecting the interceptors, but then I'd have three separate Session instances, which I believe triples the dispatch queues, right?

I'm aware of request level interceptors, but then the feature developer needs to take care to attach them, which is error prone

**Comment by jshier** [comment]
Turned this into a discussion.

This will not be a feature Alamofire adopts, as we already have many ways of composing interceptors and other features.

**Reply by ursusursus** [reply]
How would I compose interceptors shared across the instancess? You mean at request level?

**Reply by jshier** [reply]
At any level. Alamofire includes a few different premade interceptors, all of which can be composed together, either manually or using something like our provided `Interceptor` type which just stores arrays of them. 

What are you building that anyone needs to do this?

**Reply by ursusursus** [reply]
As I mentioned Lets say I have for simplicity, 2 types of requests, authenticated ones and not authenticated ones.  Whats the go to solution for this? Attach the `AuthenticationInterceptor`at request level, such as `session.request(urlRequest, interceptor: interceptor)`? (Just asking as a first step so I can ellaborate)

**Reply by ursusursus** [reply]
Please let me know if you can find the time
